doc/en/backwards-compatibility.rst

```diff
@@ -7,14 +7,15 @@ Keeping backwards compatibility has a very high priority in the pytest project.
 
 With the pytest 3.0 release we introduced a clear communication scheme for when we will actually remove the old busted joint and politely ask you to use the new hotness instead, while giving you enough time to adjust your tests or raise concerns if there are valid reasons to keep deprecated functionality around.
 
-To communicate changes we are already issuing deprecation warnings, but they are not displayed by default. In pytest 3.0 we changed the default setting so that pytest deprecation warnings are displayed if not explicitly silenced (with ``--disable-pytest-warnings``).
+To communicate changes we issue deprecation warnings using a custom warning hierarchy (see :ref:`internal-warnings`). These warnings may be suppressed using the standard means: ``-W`` command-line flag or ``filterwarnings`` ini options (see :ref:`warnings`), but we suggest to use these sparingly and temporarily, and heed the warnings when possible.
 
-We will only remove deprecated functionality in major releases (e.g. if we deprecate something in 3.0 we will remove it in 4.0), and keep it around for at least two minor releases (e.g. if we deprecate something in 3.9 and 4.0 is the next release, we will not remove it in 4.0 but in 5.0).
+We will only start the removal of deprecated functionality in major releases (e.g. if we deprecate something in 3.0 we will start to remove it in 4.0), and keep it around for at least two minor releases (e.g. if we deprecate something in 3.9 and 4.0 is the next release, we start to remove it in 5.0, not in 4.0).
```

Comment:
> and keep it around for at least two minor releases

@RonnyPfannschmidt because of this requirement, I think we should after `4.0` use a different scheme than we are using now with `RemovedInPytest4Warning`.

When we deprecate something we don't know exactly when the next major release will be. This makes it hard to track which of the functionalities marked with `RemovedInPytest4Warning` *can* be removed when pytest 4.0 comes around. For example, suppose we release `3.9` and then `4.0`. We need to track mentally what has been deprecated in `3.9`, because they can't really be removed in 4.0. Worse, they will carry the `RemovedInPytest4Warning` name with them until the 5.0 release. Because this, we can't deprecate anything else from now on, because this will lock us into getting 4.0 out if we want to avoid slipping `RemovedInPytest4Warning` into the 4.X series.

If avoid this problem, I propose this solution:

We add a version parameter to ``PytestDeprecationWarning``. When we deprecate something, we always know what that next minor release will be (for example next is `3.9`). When a new major release happens, we just need to go over the deprecation warning instances and see which have been around for more than 2 minor releases, those are the ones we should turn into errors and remove.

The more we refactor the internal code and improve the API, the more I see getting a very tight and consistent deprecation process is important to avoid disrupting our users.

Please let me know your thoughts. 😁

Reply:
> To avoid this problem, here's what I propose:
> 
> We add a version parameter to PytestDeprecationWarning. When we deprecate something, we always know what the next minor release will be (for example at the moment we know the next release is 3.9). When a new major release happens, we just need to go over the deprecation warning instances and see which have been around for more than 2 minor releases, those are the ones we should turn into errors and remove later.

Very good - I think that would make it much clearer.

One suggestion: how about naming the parameter in a clearer way, that even the casual source code reader knows what it is about. Something like `deprecated_since[_version]`?

If the docstring for the deprecation would also contain the "rules" then, this would be clear as day - e.g.

```python
class PytestDeprecationWarning(PytestWarning, DeprecationWarning):
    """
    [Same stuff as before]

    Removals only happen in major releases, after a deprecation warning has been emitted for at least two minor versions.
    """
```

Don't know if this is the best way to phrase it, but I hope you get my drift.

Reply:
All around good suggestions. 👍